### PR TITLE
IBM OpsTools - add environment. Closes #1339

### LINKED ISSUE #1339: As a user i can add an environment that will discover from ibm cloud
When adding an environment there should be the option of discover from ibm cloud this will let you log into ibm cloud and then find the node files

### PR DESCRIPTION
**Pull request by lquintai**
Signed-off-by: Leonor Quintais <lquintai@uk.ibm.com>

### CODE REVIEW

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts [hunk 1]
```diff
@@ -94,78 +97,153 @@ export async function addEnvironment(): Promise<void> {
...
-            const HEALTH_CHECK: string = '/ak/api/v1/health';
-            const GET_ALL_COMPONENTS: string = '/ak/api/v1/components';
-            let url: string;
-            const userUrl: string = await UserInputUtil.showInputBox('Enter the URL of the IBM Blockchain Platform Console you want to connect to');
-            if (!userUrl) {
+            const isSaaS: string = await UserInputUtil.showQuickPickYesNo('Are you connecting to a service instance on IBM Cloud?');
+            if (!isSaaS) {
                 return;
+            } else if (isSaaS === UserInputUtil.YES) {
+                let url: string;
```
Comment (cazfletch) [inline packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:104]: can we split some of this file out into functions. The functions can be in this file but its getting to be a really big function

File: packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts [hunk 2]
```diff
@@ -128,32 +129,52 @@ export async function importNodesToEnvironment(environmentRegistryEntry: FabricE
...
+            // establish connection to Ops Tools
             try {
-                // establish connection to Ops Tools and retrieve json file
-                const credentials: string = await keytar.getPassword('blockchain-vscode-ext', environmentRegistryEntry.url);
-                const credentialsArray: string[] = credentials.split(':');
-                if (credentialsArray.length !== 3) {
-                    throw new Error(`Unable to retrieve the stored credentials`);
+                if (environmentRegistryEntry.environmentType === EnvironmentType.SAAS_OPS_TOOLS_ENVIRONMENT) {
+                    const cloudAccountExtension: vscode.Extension<any> = vscode.extensions.getExtension( 'IBM.ibmcloud-account' );
+                    if ( !cloudAccountExtension.isActive ) {
```
Comment (cazfletch) [inline packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts:141]: Can we split this code out into a separate file as it seems duplicated, also why is there less log in code here

File: packages/blockchain-extension/package.json [hunk 3]
```diff
@@ -1504,7 +1505,8 @@
             ".pid",
             ".port",
             ".feature",
-            ".xml"
+            ".xml",
+            ".ts"
```
Comment (cazfletch) [inline packages/blockchain-extension/package.json:1509]: we shouldn't be ignoring .ts files in the licence checker as all typescript files should have a licence

Comment (lquintai) [reply packages/blockchain-extension/package.json]: indeed, modified license format (it's the same license)

File: packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts (hunk 2 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts]: done. Login code is the same, but more work is needed to get the console endpoint.

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts (hunk 1 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts]: done.

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts [hunk 4]
```diff
@@ -286,4 +241,128 @@ export async function addEnvironment(): Promise<void> {
...
+        const ibpResources: IBlockchainQuickPickItem<string>[] = [];
+        const accessToken: string = await ExtensionsInteractionUtil.cloudAccountGetAccessToken(true);
+        if (!accessToken) {
+            return;
+        }
+        const requestOptions: any = { headers: { Authorization: `Bearer ${accessToken}` } };
+        const baseUrl: string = 'https://resource-controller.cloud.ibm.com';
+        let resourcesUrl: string = `${baseUrl}/v2/resource_instances`;
+        while (resourcesUrl) {
+            const response: any = await Axios.get(resourcesUrl.toString(), requestOptions);
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:326]: Is there a reason / a scenario when resourcesUrl is not a string?

File: packages/blockchain-extension/extension/cloud-account-api.d.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * limitations under the License.
+*/
+
+import { URL } from "url";
+
+/**
+ * The external interface exposed by the IBM Cloud Account extension
+ * for use by other modules to interact with IBM Cloud APIs.
+ */
+export interface CloudAccountApi {
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/cloud-account-api.d.ts:21]: Might be worth us creating an `interfaces` directory and putting this file in there - just in case we add more in the future.

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts [hunk 6]
```diff
@@ -286,4 +241,128 @@ export async function addEnvironment(): Promise<void> {
...
+
+            for (const resource of response.data.resources) {
+                if ( resource.resource_plan_id === 'blockchain-standard' ) {
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:329]: What is the `resource_plan_id`? Does it come from IBP? Do we not want to support other resource_plan_id's?

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts [hunk 7]
```diff
@@ -286,4 +241,128 @@ export async function addEnvironment(): Promise<void> {
...
+            }
+        }
+
+        let chosenIbp: IBlockchainQuickPickItem<any>;
+        if (ibpResources.length === 0) {
+            throw new Error('There are no IBM Blockchain Platform resources associated with the chosen account');
+        } else if (ibpResources.length === 1) {
+            chosenIbp = ibpResources[0];
+        } else {
+            chosenIbp = await UserInputUtil.showQuickPickItem('Please select the IBM Blockchain Platform to connect to', ibpResources) as IBlockchainQuickPickItem<any>;
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:350]: Is this message clear enough for a user - would `Please select the IBM Blockchain Platform resource to connect to` be clearer?

File: packages/blockchain-extension/extension/util/ExtensionsInteractionUtil.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            hasAccount = await cloudAccount.accountSelected();
+            if (!hasAccount) {
+                // If not logged in, this will first ask the user to login, and then to select the account.
+                await vscode.commands.executeCommand('ibmcloud-account.selectAccount');
+            }
+        }
+
+        const isLoggedIn: boolean = await cloudAccount.loggedIn();
+        if (!isLoggedIn) {
+            return; // FIXME - evaluate if an error should be thrown after changes to cloud extension - tell the user we can't proceed without being logged in
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/util/ExtensionsInteractionUtil.ts:45]: Will this be fixed in another PR?

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts (hunk 6 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:329]: It's what identifies the resource as an IBM Blockchain platform, as opposed to any other resource in IBM cloud. We don't want to support any other.

File: packages/blockchain-extension/extension/util/ExtensionsInteractionUtil.ts (hunk 8 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/util/ExtensionsInteractionUtil.ts:45]: All the current FIXME will be addressed once there is a decision regarding the cloud login extension modifications I proposed.

File: packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts [hunk 9]
```diff
@@ -128,32 +129,44 @@ export async function importNodesToEnvironment(environmentRegistryEntry: FabricE
...
+                    }
+                    requestOptions = { headers: { Authorization: `Bearer ${accessToken}` } };
+                } else {
+                    const keytar: any = ModuleUtil.getCoreNodeModule('keytar');
+                    if (!keytar) {
+                        throw new Error('Error importing the keytar module');
+                    }
+
+                    const credentials: string = await keytar.getPassword('blockchain-vscode-ext', environmentRegistryEntry.url);
+                    const credentialsArray: string[] = credentials.split(':');
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts:152]: Just wondering where you've done 'splits', if its worth writing a comment that says what you're expecting the output to be? Might make it easier to debug if Cloud changes the values/format.

E.g. `const credentialsArray: string[] = credentials.split(':'); // Format ['hello', 'world', 'somethingElse']`

Don't know - what do you think?

File: packages/blockchain-extension/extension/cloud-account-api.d.ts (hunk 5 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/cloud-account-api.d.ts:21]: Will do.

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts (hunk 4 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:326]: Good catch, I think by then I enforced type already.

File: packages/blockchain-extension/package.json [hunk 10]
```diff
@@ -1341,7 +1341,7 @@
         "pretest": "npm run updatePackageJSON",
         "test": "node ./build/test/runTests.js",
         "posttest": "remap-istanbul -i ./coverage/coverage.json -o ./coverage/coverage-remap.json && istanbul check-coverage ./coverage/coverage-remap.json",
-        "cucumber": "node ./build/cucumber/runCucumberTests.js",
+        "cucumber": "npm run updatePackageJSON && node ./build/cucumber/runCucumberTests.js",
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/package.json:1344]: Why do we need to do this?

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts (hunk 7 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:350]: Just checked the design and actually the message should be 'Select an IBM Blockchain Platform service instance', so I'll change it to that.

Comment (lquintai) [reply packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:350]: Also changed the wording of the previous error to say 'instances' instead of 'resources'

File: packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts (hunk 9 above)
Comment (lquintai) [reply packages/blockchain-extension/extension/commands/importNodesToEnvironmentCommand.ts:152]: The format is decided by us in 'addEnvironmentCommand', so our extension is in control - there will be no changes unless we make them. Happy to add a comment for readability though.

File: packages/blockchain-extension/package.json (hunk 10 above)
Comment (lquintai) [reply packages/blockchain-extension/package.json:1344]: Ah.. Tests failed in azure, and it looked like the changes I made modified activation events slightly. Maybe something to do with activating the new extension dependency? Talked with Caroline, who's had a similar problem, and the suggestion was to updatePackageJSON just before running the cucumber tests, to guarantee activation events happen in the correct order and therefore tests start with a clean slate. This is the solution used in V2.

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts [hunk 11]
```diff
@@ -286,4 +241,128 @@ export async function addEnvironment(): Promise<void> {
...
+                if ( resource.resource_plan_id === 'blockchain-standard' ) {
+                    ibpResources.push({
+                        label: resource.name,
+                        description: resource.guid,
+                        data: resource
+                    });
+                }
+            }
+            if (response.data.next_url) {
+                resourcesUrl = `${baseUrl}${response.data.next_url.toString()}`;
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:338]: Do we need to call .toString() here?

File: packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts [hunk 12]
```diff
@@ -286,4 +241,128 @@ export async function addEnvironment(): Promise<void> {
...
+            chosenIbp = await UserInputUtil.showQuickPickItem('Select an IBM Blockchain Platform service instance', ibpResources) as IBlockchainQuickPickItem<any>;
+            if (!chosenIbp) {
+                return;
+            }
+        }
+
+        // get API endpoint
+        const dashboardUrl: URL = new URL(chosenIbp.data.dashboard_url);
+        const encodedCrn: string = encodeURIComponent(chosenIbp.data.crn);
+        const consoleStatus: any = await Axios.get(`${dashboardUrl.origin.toString()}/api/alternative-auth/resources/${encodedCrn}/optools`, requestOptions);
```
Comment (Jakeeyturner) [inline packages/blockchain-extension/extension/commands/addEnvironmentCommand.ts:359]: Do we need to call .toString() here as well?
